### PR TITLE
einsum 

### PR DESCRIPTION
**Pull request by jcmgray**
This is a general einsum implementation (#31) for the COO format. An implementation was provided in #31 but it doesn't seem to work for all inputs. The implementation here passes all of `opt_einsum`'s test suite edges cases, and is also reasonably performant (e.g. same order of magnitude speed as `tensordot` for cases where they overlap in functionality, sometimes faster). But certainly not the most optimized one could do. One would also probably want to use `opt_einsum` for more than 2 terms generally, but it's quite useful to have a reference, non-optimized form.

### Maybe later:
- [ ] dispatch to tensordot when compatible?
- [ ] benchmarks

### DONE
- [x] shape checks
- [x] main docs?
- [x] main tests
- [x] input checks (such as default fill value?)
- [x] decide what to do with mixed input types?
- [x] edge case tests such as mixed input types?
- [x] docstrings
- [x] decide what to do in the scalar output case: return scalar or sparse? : numpy returns scalar, this is generally convenient

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Thanks for the PR! It's much appreciated. However, I won't have much time to review before the weekend.

Assuming you're willing, please go ahead and work on it. Based on your comments in #31 and here, this is a very welcome change.

File: sparse/_common.py [hunk 1]
```diff
@@ -1211,6 +1212,142 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
...
+        The array to perform the einsum on.
+
+    Returns
+    -------
+    output : SparseArray
+    """
+    operand = asCOO(operand)
+
+    if lhs == rhs:
+        # XXX: take copy?
```
Comment (hameerabbasi) [inline sparse/_common.py:1235]: No need to take a copy -- `SparseArrays` are assumed immutable.

File: sparse/tests/test_einsum.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    d = 4
+
+    terms = subscripts.split("->")[0].split(",")
+    arrays = [sparse.random((d,) * len(term), density=density) for term in terms]
+    sparse_out = sparse.einsum(subscripts, *arrays)
+    numpy_out = np.einsum(subscripts, *(s.todense() for s in arrays))
+
+    if hasattr(sparse_out, "todense"):
+        # scalar output is not a sparse array
+        sparse_out = sparse_out.todense()
```
Comment (hameerabbasi) [inline sparse/tests/test_einsum.py:93]:
```suggestion
    if hasattr(sparse_out, "todense"):
        # scalar output is not a sparse array
        sparse_out = sparse_out.todense()
```

This should be done inside the main `einsum` function -- The convention is to return `SparseArray` instead of `ndarray` (including 0D), but a scalar for a scalar.

File: sparse/_common.py [hunk 3]
```diff
@@ -1211,6 +1212,142 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
...
+        new_data = operand.data[selector]
+    else:
+        new_coords = operand.coords[perm]
+        new_data = operand.data
+
+    if not rhs:
+        # scalar output - match numpy behaviour by not wrapping as array
+        return new_data.sum()
+
+    return operand.__class__(new_coords, new_data, shape=new_shape, has_duplicates=True)
```
Comment (hameerabbasi) [inline sparse/_common.py:1272]: Why not use `COO` directly?

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
A couple of suggested changes -- The base logic looks good.

**Comment by hameerabbasi** [comment]
> * main docs?

Yes, documentation would be useful, if only a section in the usage page.



> * input checks (such as default fill value?)

Shape checks and zero fill value are a must.



> * decide what to do with mixed input types?

General convention is to return `COO` if mixed, but the same type if others.



> * edge case tests such as mixed input types?

Yes, please. I've noticed bugs sometimes pop up in this area.

File: sparse/_common.py (hunk 3 above)
Comment (jcmgray) [reply sparse/_common.py:1272]: It was just to avoid the circular import error of putting `from .coo import COO` in `_common.py`,but suggestions welcome.

File: sparse/tests/test_einsum.py (hunk 2 above)
Comment (jcmgray) [reply sparse/tests/test_einsum.py:93]: That convention should already happen - possibly could be clearer, but this line just converts non-scalar (i.e. sparse) arrays to dense for the test comparison.

**Comment by jcmgray** [comment]
> Yes, documentation would be useful, if only a section in the usage page.

I've mentioned it alongside tensordot, but let me know if there should more extensive docs. I'm partly assuming the implicit "it's numpy.einsum but for sparse arrays" is enough.

> Shape checks and zero fill value are a must.
> General convention is to return COO if mixed, but the same type if others.
> Yes, please. I've noticed bugs sometimes pop up in this area.

I've added various input and shape checks. The logic of what format to output is already largely handled by the `multiply(*prepared_operands)`. The only non-trivial thing is that `DOK` didn't support `reshape`, and it was cleaner to add it rather than add logic about converting to COO and back etc. But maybe `DOK.reshape` was intentionally not implemented?

File: sparse/_common.py (hunk 3 above)
Comment (hameerabbasi) [reply sparse/_common.py]: Well, operand may not be `COO`. I'd instead import inside the function.

**Comment by hameerabbasi** [comment]
> But maybe `DOK.reshape` was intentionally not implemented?

Definitely useful to have this method, so long as it's tested.

**Comment by jcmgray** [comment]
@hameerabbasi, I've added various `DOK.reshape` tests. From my end I think this is good to go now.

**Comment by hameerabbasi** [comment]
Waiting on CI to turn green.

**Review by hameerabbasi — APPROVED** [review]

**Comment by hameerabbasi** [comment]
Just formatting errors from what I can see, run black. 😅

**Comment by jcmgray** [comment]
Aha a sneaky single quote, done.

**Comment by hameerabbasi** [comment]
Ah, still a bunch of black fixes in other files. 😅 I'll made a quick fix PR soon and then re-run the CI on and merge yours.

**Comment by jcmgray** [comment]
I see, I did run `black sparse/` and it didn't change anything else, maybe there is a particular config it needs to run with?

**Comment by hameerabbasi** [comment]
Just run `black .` with the latest black version.

**Comment by hameerabbasi** [comment]
Thanks, @jcmgray!

**Comment by jcmgray** [comment]
Nice, thanks for reviewing!
